# <center>Семинар №1<center>
### <left>Выполнил: Поверинов И.В.<left>
### <left>Группа: ИДБ-22-10<left>

Вам представлен датасет cars.csv с объявлениями о продаже в Германии на 2022 год. Необходимо осмотреть первично данные и выполнить их предобработку, также сделать соответствующие выводы по полученной информации

### Обзор данных

* `mileage` — Пройденное расстояние на автомобиле
* `make` — Марка автомобиля
* `model` — Модель автомобиля
* `fuel` — Тип топлива
* `gear` — Тип коробки передач
* `offerType` — Тип предложения, то есть «Новые», «Демонстрационные» машины и т.д.
* `price` — Цена продажи автомобиля в объявлении (в долларах)
* `hp` — Количество лошадиных сил
* `year` — Год производства автомобиля

Импортируйте необходимые библиотеки

In [1]:
import pandas as pd

### Загрузка данных

1. Загрузите необходимый датасет в переменную

In [2]:
data = pd.read_csv('cars.csv')

2. Проведите первичный осмотр данных при помощи известных вам методов(дополнительно можно использовать метод hist() для оценки распределений). Сделайте соответствующие выводы на основе полученной информации

In [3]:
# print(data.head())
# print(data.tail(10))
print(data.info())
#print(data.describe())
#print(data.hist())
#print(data.columns)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46405 entries, 0 to 46404
Data columns (total 9 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   mileage    46405 non-null  int64  
 1   make       46405 non-null  object 
 2   model      46262 non-null  object 
 3   fuel       46158 non-null  object 
 4   gear       46223 non-null  object 
 5   offerType  46405 non-null  object 
 6   price      44872 non-null  float64
 7   hp         46376 non-null  float64
 8   year       46405 non-null  int64  
dtypes: float64(2), int64(2), object(5)
memory usage: 3.2+ MB
None


Выводы:
*   В датафрейме 9 столбцов
*   пять из них имеют тип object (аналогия string из Python)
*   два столбца имеют тип Float64
*   еще два столбца имеют тип Int64
*   обьем занимаемый датафреймом 3.2+ MB
*   всего в датафрейме 46405 строк
*   в столбцах model, fuel, gear, price и hp есть пропуски


3. Переименуйте столбец `offerType` под стандартное название

In [4]:
data = data.rename(columns={'offerType': 'offer_type'})
print(data.columns)

Index(['mileage', 'make', 'model', 'fuel', 'gear', 'offer_type', 'price', 'hp',
       'year'],
      dtype='object')


### Работа с дубликатами данных

4. Необходимо вывести на экран количество дублированных строк в данных, если таковые имеются, а также избавиться от них. Не забудьте сбросить индекс в датафрейме при необходимости.

In [5]:

print(f"Количество дубликатов: {data.duplicated().sum()} ")
cleanData = data.drop_duplicates()
cleanData = cleanData.reset_index(drop = True)
print(cleanData.head())
print(f"Количество дубликатов: {cleanData.duplicated().sum()}")


Количество дубликатов: 2003 
   mileage        make   model      fuel    gear offer_type   price     hp  \
0   235000         BMW     316    Diesel  Manual       Used  6800.0  116.0   
1    92800  Volkswagen    Golf  Gasoline  Manual       Used  6877.0  122.0   
2   149300        SEAT    Exeo  Gasoline  Manual       Used  6900.0  160.0   
3    96200     Renault  Megane  Gasoline  Manual       Used  6950.0  110.0   
4   156000     Peugeot     308  Gasoline  Manual       Used  6950.0  156.0   

   year  
0  2011  
1  2011  
2  2011  
3  2011  
4  2011  
Количество дубликатов: 0


### Работа с пропусками

5. Выведите на экран количество пропусков в каждом столбце, опишите их, почему они могут возникнуть и как их возможно исправить

In [6]:
print(cleanData.isnull().sum())

mileage          0
make             0
model          137
fuel           247
gear           175
offer_type       0
price         1520
hp              24
year             0
dtype: int64


Причина возникновения пропусков может быть совершенно любого характера: простые технические или случайные ошибки, неполная информация датасета и прочее ( причин много )

Решить эту проблему, можно решить несколькими способами, например, вставить во все пропуски Unknown XD, а на самом деле, использование модуса, медианы или предыдущего значения (или близкого, исходя из анализа значения) помогут нам заполнить или удалить пропуски. И всегда есть М.О., которое поможет нам решить эту проблему эффективно.


6. Заполните пропуски в `model`,`fuel`,`gear`  
   <i>Напоминание: пропуски нужно заполнить <b>качественно</b> (т.е. не удалить и не заполнить значением 'Unknown' все NaN)</i>

In [7]:
import numpy as np
def missing_values(column):
    dtype = column.dtype
    if len(column) == 0:
        return column
    if dtype == 'object':
        return column.fillna(column.mode().iloc[0])
    elif pd.api.types.is_numeric_dtype(dtype):
        return column.fillna(column.median())
    else:
        return column.interpolate(method='linear')

## print(cleanData['model'].value_counts())
## print(cleanData['fuel'].value_counts())

cleanData['model'] = missing_values(cleanData['model'])
cleanData['fuel'] = missing_values(cleanData['fuel'])
cleanData['gear'] = missing_values(cleanData['gear'])

print(cleanData.isnull().sum())
# print(cleanData.head(25))

## print(cleanData['model'].value_counts())
## print(cleanData['fuel'].value_counts()) -- для проверки появились ли новые уникалыьне значения ( не появились )

mileage          0
make             0
model            0
fuel             0
gear             0
offer_type       0
price         1520
hp              24
year             0
dtype: int64


Приходится заполнять пропуски только так( метод mode()), ибо нет никакой СВЯЗИ между gear,model,fuel. Их связать можно только с price, который мы почистим только в следующем задании, и то сомнительно, ведь высокий price != АКПП, даже знание марки, коробки передач и типа топлива не даст нам информации о модели.

7. Удалите оставшиеся пропуски в столбцах `hp` и `price`.

In [8]:
cleanData = cleanData.dropna(subset=['hp', 'price'])

Проверьте, все ли пропуски точно удалились из датафрейма

In [9]:
print(cleanData.isnull().sum())

mileage       0
make          0
model         0
fuel          0
gear          0
offer_type    0
price         0
hp            0
year          0
dtype: int64


8. Оцените количество уникальных марок автомобилей (производителей) с помощью метода unique() и выведите их количество числом (пожалуйста, не считайте вручную)

In [10]:
uniqueModel = cleanData['make'].unique()
print(len(uniqueModel))

76


9. Выведите топ 20 марок автомобилей (производителей) по количеству объявлений с помощью метода value_counts()

In [11]:
print(cleanData['make'].value_counts().head(20))

make
Volkswagen       6510
Opel             4455
Ford             4114
Skoda            2665
Renault          2610
Audi             2479
BMW              2296
Mercedes-Benz    2217
SEAT             1773
Hyundai          1688
Fiat             1435
Peugeot          1136
Toyota           1106
Kia               937
smart             907
Citroen           888
Volvo             745
Nissan            705
Mazda             677
Dacia             630
Name: count, dtype: int64


### Группировка данных

10. Сгруппируйте данные по столбцу `make`, выведите для каждого производителя среднюю стоимость их автомобиля, а также среднее число лошадиных сил и отсортируйте по убыванию столбца `price`. Отобразите только топ 20 самых дорогих производителей. Напишите вывод, всё ли в порядке с полученными данными или Вас что-то смущает

In [12]:
groupedData = cleanData.groupby('make')
result = groupedData.agg({'price': 'mean','hp': 'mean'})
top = result.sort_values('price', ascending=False).head(20)
print(top)

                     price          hp
make                                  
Maybach      450479.333333  543.000000
Ferrari      324028.181818  693.909091
Lamborghini  305698.777778  690.111111
McLaren      212474.000000  654.300000
Rolls-Royce  194900.000000  591.500000
Bentley      192123.307692  546.615385
Aston        163442.960000  536.200000
Corvette     110160.000000  687.000000
Porsche       89858.615044  412.340708
FISKER        69900.000000  408.000000
Maserati      69118.818182  376.363636
Alpina        66046.300000  443.000000
Polestar      64150.000000  408.000000
Tesla         62866.681818  490.500000
Morgan        61950.000000  177.000000
Alpine        53783.000000  244.800000
Land          51287.433121  246.070064
RAM           47945.000000  401.000000
Cadillac      45032.142857  348.285714
Dodge         44968.818182  395.636364


Разница между первым из топа и последним слишком большая, это меня и смущает, похоже на неполный датасет.

У феррари слишком большой показать hp, хм....

Но если закрыть на это глаза, вроде бы все в порядке.

11. Создайте новый столбец из `make` и `model` и выведите топ 10 самый дорогих и дешевых автомобилей с помощью средней стоимости

In [13]:
cleanData.loc[:, 'rich'] = cleanData['make'].astype(str) + ' ' + cleanData['model'].astype(str)

average = cleanData.groupby('rich')['price'].mean()
top = average.nlargest(10)
bottom = average.nsmallest(10)

def format(num):
    return f"{num:.2f}"

Ftop = top.apply(lambda x: format(x))
Fbottom = bottom.apply(lambda x: format(x))

print("Топ 10 самых дорогих автомобилей:")
print(Ftop)
print("\nТоп 10 самых дешевых автомобилей:")
print(Fbottom)


Топ 10 самых дорогих автомобилей:
rich
Ferrari F12              1199900.00
Mercedes-Benz S 650       717078.00
Maybach Pullman           608439.00
Ferrari 812               439900.00
Aston Martin              398000.00
Lamborghini Aventador     371423.17
McLaren 675LT             297900.00
Ferrari F8 Tributo        294416.67
Bentley Flying Spur       251560.00
McLaren 720S              251145.00
Name: price, dtype: object

Топ 10 самых дешевых автомобилей:
rich
Lada Priora              1190.00
Brilliance BS4           1500.00
SEAT Panda               1650.00
Others Landwind          1850.00
Volvo Golf               2499.00
Trailer-Anhänger Golf    2750.00
Hyundai Getz             2800.00
Volkswagen Fox           3011.18
Peugeot 206              3363.00
Fiat Linea               3400.00
Name: price, dtype: object


12. Сделайте общий вывод о проделанной работе

Вывод о работе:
*   проанализировали датасет cars.csv
*   посмотрели и почистили дубликаты
*   посмотрели пропуски и попытались их устранить
*   поработали с данными, и проанализировали их с помощью группировки

Также столкнулись с интересным заданием: "Качественно заполнить пропуски". В котором приходилось заполнять пропуски моделей, топлива и коробки передач Автомобилей.

P.S. Оказывается марка автомобиля others – это не “какие-то еще” марки, а реальные марки корейских авто)))
